File: src/client/java/com/somefrills/features/misc/glowmob/chestui/UIInventory.java
package com.somefrills.features.misc.glowmob.chestui;

import net.minecraft.entity.player.PlayerEntity;
import net.minecraft.inventory.Inventory;
import net.minecraft.item.ItemStack;
import net.minecraft.util.collection.DefaultedList;

public class UIInventory implements Inventory {
    private final int size;
    private final DefaultedList<ItemStack> stacks;

    public UIInventory(int size) {
        this.size = size;
        this.stacks = DefaultedList.ofSize(size, ItemStack.EMPTY);
    }

    @Override
    public int size() {
        return this.size;
    }

    @Override
    public boolean isEmpty() {
        for (ItemStack stack : this.stacks) {
            if (!stack.isEmpty()) {
                return false;
            }
        }
        return true;
    }

    @Override
    public ItemStack getStack(int slot) {
        if (slot < 0 || slot >= this.size) {
            return ItemStack.EMPTY;
        }
        return this.stacks.get(slot);
    }

    @Override
    public ItemStack removeStack(int slot, int amount) {
        if (slot < 0 || slot >= this.size) {
            return ItemStack.EMPTY;
        }
        ItemStack stack = this.stacks.get(slot);
        if (stack.isEmpty()) {
            return ItemStack.EMPTY;
        }
        ItemStack result = stack.split(amount);
        if (stack.isEmpty()) {
            this.stacks.set(slot, ItemStack.EMPTY);
        }
        this.markDirty();
        return result;
    }

    @Override
    public ItemStack removeStack(int slot) {
        if (slot < 0 || slot >= this.size) {
            return ItemStack.EMPTY;
        }
        ItemStack stack = this.stacks.get(slot);
        this.stacks.set(slot, ItemStack.EMPTY);
        if (!stack.isEmpty()) {
            this.markDirty();
        }
        return stack;
    }

    @Override
    public void setStack(int slot, ItemStack stack) {
        if (slot >= 0 && slot < this.size) {
            this.stacks.set(slot, stack);
            this.markDirty();
        }
    }

    @Override
    public void markDirty() {
    }

    @Override
    public boolean canPlayerUse(PlayerEntity player) {
        return true;
    }

    @Override
    public void clear() {
        this.stacks.clear();
        this.stacks.addAll(DefaultedList.ofSize(this.size, ItemStack.EMPTY));
        this.markDirty();
    }

    public DefaultedList<ItemStack> getStacks() {
        return this.stacks;
    }
}

File: src/main/kotlin/com/somefrills/features/misc/glowmob/chestui/UIInventory.kt
package com.somefrills.features.misc.glowmob.chestui

import net.minecraft.core.NonNullList
import net.minecraft.world.Container
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack

class UIInventory(private val size: Int) : Container {

    private val stacks: NonNullList<ItemStack> =
        NonNullList.withSize(size, ItemStack.EMPTY)

    override fun getContainerSize(): Int {
        return size
    }

    override fun isEmpty(): Boolean {
        for (stack in stacks) {
            if (!stack.isEmpty) {
                return false
            }
        }
        return true
    }

    override fun getItem(slot: Int): ItemStack {
        if (slot !in 0..<size) {
            return ItemStack.EMPTY
        }
        return stacks[slot]
    }

    override fun removeItem(slot: Int, amount: Int): ItemStack {
        if (slot !in 0..<size) {
            return ItemStack.EMPTY
        }

        val stack = stacks[slot]
        if (stack.isEmpty) {
            return ItemStack.EMPTY
        }

        val result = stack.split(amount)

        if (stack.isEmpty) {
            stacks[slot] = ItemStack.EMPTY
        }

        setChanged()
        return result
    }

    override fun removeItemNoUpdate(slot: Int): ItemStack {
        if (slot !in 0..<size) {
            return ItemStack.EMPTY
        }

        val stack = stacks[slot]
        stacks[slot] = ItemStack.EMPTY

        if (!stack.isEmpty) {
            setChanged()
        }

        return stack
    }

    override fun setItem(slot: Int, stack: ItemStack) {
        if (slot in 0..<size) {
            stacks[slot] = stack
            setChanged()
        }
    }

    override fun setChanged() {
    }

    override fun stillValid(player: Player): Boolean {
        return true
    }

    override fun clearContent() {
        stacks.clear()
        setChanged()
    }
}